### PR TITLE
Keep track of `ViewMeta` with symbolic inputs.

### PR DESCRIPTION
**Pull request by ysiraichi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #125876

Fix: #125387

This PR helps keep track of whether an instantiated `ViewMeta` has symbolic values as
input or not. This is used for checking whether we use the AOTAutograd `ViewMeta`-replay
execution path, e.g. it doesn't support tensors that have `ViewMeta` with symbolic inputs.

In summary, the changes are:

- Add the field `ViewMeta::has_symbolic_inputs` and make it a required constructor
parameter
- Add the field `FunctionalTensorWrapper::is_symbolic_` and the method
`FunctionalTensorWrapper::maybe_mark_symbolic`
    - Marks a `FunctionalTensorWrapper` as symbolic iff any of its `ViewMeta` have
    symbolic inputs
- Add the plumbing of `FunctionalTensorWrapper::is_symbolic` to the Python API
- Codegen the computation of `ViewMeta::has_symbolic_inputs` for each view operation
- Use the AOTAutograd `ViewMeta`-replay path if:
    - `target_functional_tensor` is not `None`; and
    - `target_functional_tensor` is not symbolic (instead of using a functorch config)

cc @bdhirsh @miladm @lezcano

### CODE REVIEW

File: aten/src/ATen/FunctionalStorageImpl.h [hunk 1]
```diff
@@ -31,14 +31,16 @@ struct ViewMeta {
   ViewMeta(
       std::function<Tensor(const Tensor&, int64_t)> forward,
       std::function<Tensor(const Tensor&, const Tensor&, int64_t)> reverse,
+      bool has_symbolic_inputs,
```
Comment (ezyang) [inline aten/src/ATen/FunctionalStorageImpl.h:34]: So what, this should be true if the std::function closes over SymInts, or something? Spell it out.

File: torch/csrc/autograd/python_torch_functions_manual.cpp [hunk 2]
```diff
@@ -673,7 +696,7 @@ static PyObject* THPVariable__functionalize_apply_view_metas(
       {"_functionalize_apply_view_metas(Tensor tensor, Tensor base)"},
       /*traceable=*/true);
 
-  ParsedArgs<4> parsed_args;
+  ParsedArgs<2> parsed_args;
```
Comment (ezyang) [inline torch/csrc/autograd/python_torch_functions_manual.cpp:699]: whoops

**Review by ezyang — APPROVED** [review]
Having to do all this codegen for computing if there are free symbolic values is a bit wrought but I guess there isn't really any other way to do it unless we get in the business of explicitly reifying ViewMeta. I sort of predict that we might dumping all this when we get fully reified view functions, but I'm not one to stop a working PR.

**Comment by ydshieh** [comment]
Thanks for the fix!

**Pull request closed without merging**